Add serverpath preference

BrianBlade opened the pull request:
Adds a configurable _path_ to the server URL. Nice to have in some scenarios like reverse proxying.

tananaev commented:
If we want to allow users to specify path, it's probably better to just have a single URL parameter instead of all separate values (host, port, path etc).

BrianBlade commented:
I inferred a deliberate design pattern from the effort put into separating domain/address, port and scheme (aka encryption), thus added the path as another distinct option. Combining these in a single Server URL would be great for tech-savvy users but might be more difficult for users that do not care about protocol scheme and portnumber. Also, having the port as separate preference seems reasonable due to the default value. I am incertain regarding the encryption preference, however. 

Maybe we should just broaden the address preference to allow for scheme, address and path while disregarding port and sync the entered scheme with the encryption preference?

tananaev commented:
Most users don't set up encryption, so probably default "http" prefix will work for them just fine. Port is always a problem. They are always trying to change it even in a separate field. I am thinking that it would be nice to have some special warning message if user selects port 8082. Otherwise default port 5055 in the URL should just fine.

I think most users are familiar with URLs. It might be even easier for them to put whole URL than separate fields. I often hear complaint that they can't save server address. Turns out they are trying to put whole URL there instead of just IP or domain name.

BrianBlade commented:
https://github.com/tananaev/traccar-client-android/pull/269/commits/1ecd8ed454f9a2b3a194e00f2bba92c1e9c251e7 replaces separate settings of host, port & encryption with a single _Server URL_ preference. 
Open Points:
 * Currently a toast message is used as a warning if port was set to 8082, but I could not find a proper wording for the message.
* A safe upgrade path is missing - users would lose their settings on app upgrade. Not sure how to prevent that.
* I kept a check of the entered hostname against your `DOMAIN_NAME` pattern. I think this check could be omitted?

tananaev commented:
Two big things that need fixing:

1. Why do you need parse URL and format it again? You just need to verify that it's a valid URL and check port.

2. You should not change any localisations other than default English.

BrianBlade commented:
>  Why do you need parse URL and format it again? You just need to verify that it's a valid URL and check port.

Several reasons why I went that way:
* get rid of user provided credentials, query-string or fragment-identifier
* replace otherwise 'valid' protocols  like `ftp://` with `http://`
* add the default port so the user can at least see what is used if no port was manually set

Regarding your 2nd point: Sorry, I thought obsolete strings should be removed from locales as well, will fix it later on.

tananaev commented:
I think all that is unnecessary. Why would you replace "ftp" with "http"? If anything, we should disallow user entering anything other than "http(s)", but even that is too much validation IMHO.

What we should definitely implement instead is migration from previous version.

BrianBlade commented:
Finally got around to adjust the commit to your recommendations. Not sure about the migration, though. The approach works but is rather naive and simply checks for existence of a `port` preference on creation.

tananaev commented on app/src/main/java/org/traccar/client/MainActivity.java:
```diff
@@ -65,12 +68,12 @@
         }
 
         sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
+        migratePreferencesIfNecessary(sharedPreferences);
```
I would rename it to something like `migrateLegacyPreferences` to make it more clear what it actually does.

tananaev commented on app/src/main/java/org/traccar/client/MainActivity.java:
```diff
@@ -307,4 +291,47 @@ public void onRequestPermissionsResult(int requestCode, String[] permissions, in
+        if (userUrl == null || userUrl.trim().length() == 0) {
+            preference.setText(getString(R.string.settings_url_default_value));
+            preference.setSummary(R.string.settings_url_summary);
+            findPreference(KEY_STATUS).setEnabled(true);
```
You should not change status. Only user changes status.

BrianBlade replied:
If an invalid URL had been entered before, Status preference was greyed out. Here it is re-enabled since default URL is obviously valid.

tananaev replied:
That's not how validation worked before, and it still works differently for other parameters. I think we should be consistent and use original way of validating parameters.

tananaev replied:
That's not how validation worked before, and it still works differently for other parameters. I think we should be consistent and use original way of validating parameters.

tananaev commented on app/src/main/java/org/traccar/client/MainActivity.java:
```diff
@@ -80,30 +83,13 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
-                    }
-                }
-                return false;
+                return (newValue != null) && setServerURL(newValue.toString());
```
You should not "set" anything here. It's only validation.

BrianBlade replied:
I think its a bit more than validation. If an invalid URL is entered, the user is notified and the Status preference is greyed out - however, the invalid URL is kept as URL preference's value. Otherwise the user needed to re-enter the whole URL every time.
Also, if the URL preference is left empty, the default URL is used instead.
The method could of course be renamed, if thats your concern.

tananaev replied:
I think you are overcomplicating it. URL is something user doesn't enter every day. It's usually configured once. Also, there will be a default value for reference, so it's unlikely that user would enter completely invalid value.

tananaev commented on app/src/main/java/org/traccar/client/MainActivity.java:
```diff
@@ -307,4 +291,47 @@ public void onRequestPermissionsResult(int requestCode, String[] permissions, in
+            editor.remove("port");
+            editor.remove("address");
+            editor.remove("secure");
+            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.GINGERBREAD) {
```
Just use `commit`.

tananaev commented on app/src/main/java/org/traccar/client/ProtocolFormatter.java:
```diff
@@ -19,28 +19,32 @@
-        if (alarm != null) {
-            builder.appendQueryParameter("alarm", alarm);
+        if (alarm != null && alarm.trim().length() != 0) {
+            serverUrl = serverUrl.buildUpon().appendQueryParameter("alarm", alarm).build();
```
Why?

tananaev commented on app/src/main/java/org/traccar/client/ProtocolFormatter.java:
```diff
@@ -19,28 +19,32 @@
-
-        return builder.build().toString();
+        int port = serverUrl.getPort();
+        if (port < 0 || port > 65535) {
```
This should be in the preference validation, not here.

BrianBlade replied:
Since you asked not to rewrite the userinput last time, this validation would occur in two places then?

tananaev commented on app/src/main/res/values/strings.xml:
```diff
@@ -21,21 +20,14 @@
-  <string name="error_invalid_address">Invalid server address</string>
-  <string name="error_invalid_port">Invalid server port</string>
-  <string name="error_invalid_frequency">Invalid frequency value</string>
+  <string name="settings_invalid_url_summary">Invalid server URL</string>
```
Please use `error_` prefix for both new strings and put them next to each other in the file.

tananaev commented on app/src/test/java/org/traccar/client/ProtocolFormatterTest.java:
```diff
@@ -22,10 +22,17 @@ public void testFormatRequest() throws Exception {
         position.setTime(new Date(0));
 
-        String url = ProtocolFormatter.formatRequest("localhost", 5055, false, position);
+        String url = ProtocolFormatter.formatRequest("http://localhost:5055/path", position);
```
There should be at least one test without path because that's what Traccar Client will use by default.

BrianBlade commented:
Thanks for your feedback!
I think I addressed your concerns with the latest commits. Please let me know if you have any further remarks or want me to squash the commits.

tananaev commented on app/src/main/java/org/traccar/client/MainActivity.java:
```diff
@@ -307,4 +289,38 @@ public void onRequestPermissionsResult(int requestCode, String[] permissions, in
     }
 
+    private boolean validateServerURL(String userUrl) {
+        if (userUrl == null || userUrl.trim().length() == 0) {
```
Why do we reset to the default one?

BrianBlade replied:
If someone explicitly enters "nothing" it makes more sense to use the default value than to just reject the action, imho.

tananaev replied:
I don't think we should explicitly handle this case. I feel like you are trying to overcomplicate it for no reason. There can never be null value as far as I understand it. If the value is empty, URL validation will handle it.

tananaev commented on app/src/main/java/org/traccar/client/MainActivity.java:
```diff
@@ -307,4 +289,38 @@ public void onRequestPermissionsResult(int requestCode, String[] permissions, in
+            return false;
+        }
+        int port = Uri.parse(userUrl).getPort();
+        if (URLUtil.isValidUrl(userUrl) && (port >= -1 && port <= 65535)
```
Doesn't `isValidUrl` validate port as well?

BrianBlade replied:
I thought so as well but your former comment made me check and unfortunately it does not.

tananaev replied:
OK, but why `>= -1`? I think it should be `> 0`.

BrianBlade replied:
`getPort()` returns `-1` if no port was set in which case default port `5055` will be used as per your former request

tananaev replied:
I don't remember requesting such thing. Default port for HTTP is 80 and HTTPS is 443.

I think we should only validate port if it's provided.

tananaev commented:
Merged, thanks.
